controller/sharding/sharding.go

```diff
@@ -158,6 +162,59 @@ func RoundRobinDistributionFunction(clusters clusterAccessor, replicas int) Dist
 	}
 }
 
+// ConsistentHashingWithBoundedLoadsDistributionFunction returns a DistributionFunction using an almost homogeneous distribution algorithm:
+// for a given cluster the function will return the shard number based on a consistent hashing with bounded loads algorithm.
+// This function ensures an almost homogenous distribution: each shards got assigned the fairly similar number of
+// clusters +/-10% , but with it is resilient to sharding and/or number of clusters changes.
+func ConsistentHashingWithBoundedLoadsDistributionFunction(clusters clusterAccessor, replicas int) DistributionFunction {
+	return func(c *v1alpha1.Cluster) int {
+		if replicas > 0 {
+			if c == nil { // in-cluster does not necessarly have a secret assigned. So we are receiving a nil cluster here.
+				return 0
+			}
+
+			// if Shard is manually set and the assigned value is lower than the number of replicas,
+			// then its value is returned otherwise it is the default calculated value
+			if c.Shard != nil && int(*c.Shard) < replicas {
+				return int(*c.Shard)
+			} else {
+				// if the cluster is not in the clusters list anymore, we should unassign it from any shard, so we
+				// return the reserved value of -1
+				if !slices.Contains(clusters(), c) {
+					log.Warnf("Cluster with id=%s not found in cluster map.", c.ID)
+					return -1
+				}
+				consistentHashing := createConsistentHashingWithBoundLoads(replicas)
+				clusterIndex, err := consistentHashing.Get(c.ID)
+				if err != nil {
+					log.Warnf("Cluster with id=%s not found in cluster map.", c.ID)
+					return -1
+				}
+				shard, err := strconv.Atoi(clusterIndex)
+				if err != nil {
+					log.Errorf("Consistent Hashing was supposed to return a shard index but it returned %d", err)
+					return -1
+				}
+				log.Debugf("Cluster with id=%s will be processed by shard %d", c.ID, shard)
+				return shard
```

Comment:
Actually what you want to do is to get a map with the results of the consistent hash. The hash has to be consistent across all the clusters. Changed the function to return a map where the consistent hash will be used to build the map:
```
shardIndexedByCluster := createConsistentHashingWithBoundLoads(replicas, clusters, apps)
shard, ok := shardIndexedByCluster[c.ID]
if !ok {
	log.Warnf("Cluster with id=%s not found in cluster map.", c.ID)
	return -1
}
log.Debugf("Cluster with id=%s will be processed by shard %d", c.ID, shard)
return shard
```

Reply:
Fixed based on your comments @Enclavet